boot/spring_performance.go

```diff
@@ -0,0 +1,265 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package boot
+
+import (
+	"archive/zip"
+	"fmt"
+
+	"github.com/paketo-buildpacks/libpak/sherpa"
+
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/buildpacks/libcnb"
+	"github.com/magiconair/properties"
+	"github.com/paketo-buildpacks/libpak"
+	"github.com/paketo-buildpacks/libpak/bard"
+	"github.com/paketo-buildpacks/libpak/effect"
+)
+
+type SpringPerformance struct {
+	Dependency       libpak.BuildpackDependency
+	LayerContributor libpak.LayerContributor
+	Logger           bard.Logger
+	Executor         effect.Executor
+	AppPath          string
+	Manifest         *properties.Properties
+	AotEnabled       bool
+	DoTrainingRun    bool
+	ClasspathString  string
+	ReZip            bool
+}
+
+func NewSpringPerformance(cache libpak.DependencyCache, appPath string, manifest *properties.Properties, aotEnabled bool, doTrainingRun bool, classpathString string, reZip bool) SpringPerformance {
+	contributor := libpak.NewLayerContributor("Performance", cache, libcnb.LayerTypes{
+		Build:  true,
+		Launch: true,
+	})
+	return SpringPerformance{
+		LayerContributor: contributor,
+		Executor:         effect.NewExecutor(),
+		AppPath:          appPath,
+		Manifest:         manifest,
+		AotEnabled:       aotEnabled,
+		DoTrainingRun:    doTrainingRun,
+		ClasspathString:  classpathString,
+		ReZip:            reZip,
+	}
+}
+
+func (s SpringPerformance) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
+	s.LayerContributor.Logger = s.Logger
+	layer, err := s.LayerContributor.Contribute(layer, func() (libcnb.Layer, error) {
+
+		layer.LaunchEnvironment.Default("BPL_SPRING_AOT_ENABLED", s.AotEnabled)
+
+		if !s.DoTrainingRun {
+			return layer, nil
+		}
+
+		layer.LaunchEnvironment.Default("BPL_JVM_CDS_ENABLED", s.DoTrainingRun)
+
+		// prepare the training run JVM opts
+		var trainingRunArgs []string
+
+		if s.AotEnabled {
+			trainingRunArgs = append(trainingRunArgs, "-Dspring.aot.enabled=true")
+		}
+
+		jarPath := s.AppPath
+
+		if s.ReZip {
+			jarDestDir := os.TempDir() + "/" + fmt.Sprint(time.Now().UnixMilli()) + "/jar-dest"
+			if err := os.MkdirAll(jarDestDir, 0755); err != nil {
+				return layer, fmt.Errorf("error creating temp directory for jar\n%w", err)
+			}
+			tempJarPath := filepath.Join(jarDestDir, "runner.jar")
+			if err := CreateJar(s.AppPath+"/", tempJarPath); err != nil {
+				return layer, fmt.Errorf("error recreating jar\n%w", err)
+			}
+			f, err := os.Open(tempJarPath)
+			if err != nil {
+				return layer, fmt.Errorf("error opening jar\n%w", err)
+			}
+			if err = sherpa.CopyFile(f, filepath.Join(layer.Path, "runner.jar")); err != nil {
+				return layer, fmt.Errorf("error copying jar\n%w", err)
+			}
```

Comment:
that's because the Spring Boot feature that extracts into a CDS friendly format can not be called from an unzipped Spring Boot jar; see : https://github.com/spring-projects/spring-boot/issues/38276#issuecomment-1983451808

I know you may think : "what are they doing? unzipping, rezipping, and then creating a new layout? what a waste of IO!"

But: we have to make sure we deal with a spring boot jar (+1 for unzip) , apps run slightly faster when unzipped (+1 for unzip) preferably one single path when we get to `performance` (+1 for unzip), potentially add the SCB jar into the app (+1 for unzip) - and then we need to call spring boot tooling (+1 for rezip)

And, it happens at build time, so I don't think that playing with IO is a huge concern.

Reply:
I'm really just trying to understand everything, hence all the questions. I appreciate all the detailed responses.

I'd like to know more about what Boot is doing with this extraction command. In a nutshell, it sounds like it's extracting some stuff from the standard fat JAR and putting it into a different format (the comment made it seem like libs are pulled out and a "runner" jar is created). Since we already have the files extracted, it seems like we could move some files around on disk and arrive at the new format bypassing that command. Not knowing more about what Boot is doing, there may be details missing that might make this more complicated than it's worth.

I do think we need to consider the performance here. We're talking about the user's application, so it is an unbounded number of files. We could have 20M of files, we could have 2G of files so we can't just write this off as trivial. It could be a pretty noticeable performance impact for some users. The other point to consider is caching. I don't think you're going to be able to cache this very effectively. This is operating on the user's files, so you pretty much have to do this every time you build, since user's files could change every time. That means a performance hit is going to be felt on every build, not just first builds or occasionally when cache is invalidated.

Given that, I think we have to at least consider the level of effort for implementing this more efficiently. Maybe implementing the logic being done by the extract command in Go (or some other alternative solution), without all the extra file i/o and CPU. Maybe it's too complicated and we can't justify it but it seems worth at least discussing what that command is doing more or doing a spike to get a better handle on the decision.

-----------

Another concern of mine is the state of the user's application. 

1. When you build from their source, we get the user's source code in `/workspace`. 
    a. Maven/Gradle run and provide a Spring Boot JAR. The buildpack then extracts this, removes source code, and puts the compiled code into `/workspace`.
    b. Maven/Gradle run and provide multiple JAR files. The buildpack then takes all of these JAR files, removes source code, and puts all the JARs not extracted into `/workspace`.
2a. When the user provides a Spring Boot fat JAR, pack extracts this and puts it into `/workspace`.
2b. I think it fails at the moment with executable-jar, but a user can technically provide a directory of JAR files they pre-compiled, and pack will put set them as the contents of `/workspace`. This is probably a scenario that we'll support in the future though, there's been a desire to fix executable-jar in this regard.

It's the 1b.) and 2b.) scenarios that concern me. I think neither is probably super likely for a Boot app, but how are these situations going to be handled? Are we going to fail them as unsupported? Or are they legit and we need to handle them?

If we are not intending to handle them, I think that might simplify some of the logic around when you need to zip/unzip stuff.

Reply:
> I'm really just trying to understand everything, hence all the questions. I appreciate all the detailed responses.

👍

> what Boot is doing with this extraction command. In a nutshell, it sounds like it's extracting some stuff from the standard fat JAR and putting it into a different format

yeah something like that: extracts all jars in BOOT-INF/lib into their own folder, creates a jar with a "normal" java MANIFEST and all users classes and the list of jars in `BOOT-INF/lib`

>  Maybe implementing the logic being done by the extract command in Go

well, this is how we started with @pivotal-david-osullivan - but we reverted course because we don't want the `spring-boot` buildpack to "know" about internal Spring Boot transformation. I spend a lot of time actually trying to write a META-INF/MANIFEST and I can tell you, that's not just limited lines, weird indentation or whatnot; and there's no existing Go library that handles that.

It's a Spring Boot + Java transformation - I'm afraid Go is not the proper tooling to take care of that; a broken MANIFEST could hurt the user in weird ways...

Reply:
> It's the 1b.) and 2b.) scenarios that concern me. I think neither is probably super likely for a Boot app

right, there's no way a Spring Boot app would come as a bunch of jars; like I wrote a little bit earlier, Spring Boot team keeps the output options very limited

> If we are not intending to handle them, I think that might simplify some of the logic around when you need to zip/unzip stuff.

hum... do you mean we could remove the `findSpringBootExecutableJAR` function? at least the walking part? that does not seem like a huge gain.